Improve debug logs

RobinMalfait opened the pull request:
This PR improves the debug logs for the `@tailwindcss/postcss` integration. It uses custom instrumentation to provide a nested but detailed overview of where time is spent during the build process.

The updated logs look like this:
```
[0.15ms] [@tailwindcss/postcss] src/app/geistsans_9fc57718.module.css
[0.14ms]   ↳ Quick bail check
[0.02ms] [@tailwindcss/postcss] src/app/geistsans_9fc57718.module.css
[0.01ms]   ↳ Quick bail check

[0.03ms] [@tailwindcss/postcss] src/app/geistmono_b9f59162.module.css
[0.02ms]   ↳ Quick bail check
[0.12ms] [@tailwindcss/postcss] src/app/geistmono_b9f59162.module.css
[0.11ms]   ↳ Quick bail check

[42.09ms] [@tailwindcss/postcss] src/app/globals.css
[ 0.01ms]   ↳ Quick bail check
[12.12ms]   ↳ Setup compiler
[ 0.11ms]     ↳ PostCSS AST -> Tailwind CSS AST
[11.99ms]     ↳ Create compiler
[ 0.07ms]   ↳ Register full rebuild paths
[ 0.06ms]   ↳ Setup scanner
[ 7.51ms]   ↳ Scan for candidates
[ 5.86ms]   ↳ Register dependency messages
[ 5.88ms]   ↳ Build utilities
[ 8.34ms]   ↳ Optimization
[ 0.23ms]     ↳ AST -> CSS
[ 4.20ms]     ↳ Lightning CSS
[ 3.89ms]     ↳ CSS -> PostCSS AST
[ 1.97ms]   ↳ Update PostCSS AST
```

RobinMalfait committed:
add instrumentation

RobinMalfait committed:
fix missing `console.timeEnd(…)`

If we bail early, then a `console.time('…')` was already started and
subsequent calls to `console.timeEnd('…')` will fail because the
`console.time` and `console.timeEnd` don't line up.

RobinMalfait committed:
splitup steps, and add a few more timing logs

RobinMalfait committed:
add used file to the debug logs

RobinMalfait committed:
only enable oxide logs using `DEBUG=tailwindcss:oxide` or `DEBUG=*`

RobinMalfait committed:
use custom intstrumentation

RobinMalfait committed:
add a few more logs

RobinMalfait committed:
use custom instrumentation in `@tailwindcss/cli`

philipp-spiess commented on packages/@tailwindcss-node/src/instrumentation.test.ts:
```diff
@@ -0,0 +1,57 @@
+
+  I.report((output) => {
+    expect(stripVTControlCharacters(output).replace(/\[.*\]/g, '[0.xxms]')).toMatchInlineSnapshot(`
+      "Hits:
+        Potato × 4
+
+      Timers:
+      [0.xxms] Foo
+      [0.xxms]   ↳ Bar × 100"
```
This is neat!

philipp-spiess commented on packages/@tailwindcss-postcss/src/index.ts:
```diff
@@ -161,14 +194,16 @@ function tailwindcss(opts: PluginOptions = {}): AcceptedPlugin {
 
             // Look for candidates used to generate the CSS
             context.scanner = new Scanner({ sources })
+            env.DEBUG && I.end('Setup scanner')
```
wonder if we should reassign `env.DEBUG` to a local variable so we save on the repeated property access

RobinMalfait committed:
implement `Instrumentation` as a class

This also allows us to make it disposable. This allows us to create a
new instrumentation instance using the `using` keyword, which
automatically reports the findings when it's being disposed.

RobinMalfait committed:
ensure `using` keyword is being compiled

RobinMalfait committed:
use new `Instrumentation`

RobinMalfait committed:
only report when `env.DEBUG` is true

This is only the case for when we are automatically disposing the
`Instrumentation`.

RobinMalfait committed:
use the `Instrumentation` in the `Vite` plugin

RobinMalfait committed:
use shared `DEBUG` constant

This prevents `env.DEBUG` property access over and over again

RobinMalfait committed:
fix import

RobinMalfait commented on packages/@tailwindcss-cli/src/commands/build/index.ts:
```diff
@@ -66,6 +67,9 @@ async function handleError<T>(fn: () => T): Promise<T> {
 }
 
 export async function handle(args: Result<ReturnType<typeof options>>) {
+  using I = new Instrumentation()
```
Ideally this is only created when `DEBUG` is true but I don't think that's possible

RobinMalfait replied:
If you look at the compiled output (because we can't use the native version just yet), what it essentially does is it wraps the code in a try/catch/finally and calls the dispose function in the finally block.

E.g.: https://www.typescriptlang.org/play/?#code/DYUwLgBAZg9jEF4IG8BQEJhAZzACgEoUIBfAGnQgG0BlATwFsAjGYAOgBMBLbABxmwgAuoRSUMAYxgA7bKxBtgMAOZ4A5Bxkg1BSiVT7UUAK7SJYLjIgALEMCWi0GY9i7TlEAB6JocSp7YsXEIDIA

The cool part is that you don't have to worry about all the spots where you used an early return to flush the changes to the terminal because the moment you return, the `finally` block is executed as well.

But yeah, I don't know/think you can create one conditionally because of this try/catch setup 😬

thecrypticace replied:
I think it's fine. It's like a few objects and closures being created. The overhead for this is massively dwarfed by everything else. And reporting doesn't happen unless in debug mode either so that code does nothing.

RobinMalfait commented on crates/oxide/src/lib.rs:
```diff
@@ -24,7 +24,7 @@ pub mod paths;
 
 static SHOULD_TRACE: sync::LazyLock<bool> = sync::LazyLock::new(
-    || matches!(std::env::var("DEBUG"), Ok(value) if value.eq("*") || value.eq("1") || value.eq("true") || value.contains("tailwind")),
+    || matches!(std::env::var("DEBUG"), Ok(value) if value.eq("*") || (value.contains("tailwindcss:oxide") && !value.contains("-tailwindcss:oxide"))),
```
When using `DEBUG=1` it also shows a lot of oxide related tracing logs. This change only shows them if you are using `DEBUG=*` or `DEBUG=tailwindcss:oxide` by following the `DEBUG` conventions.

RobinMalfait commented on packages/@tailwindcss-cli/src/commands/build/index.ts:
```diff
@@ -226,32 +233,36 @@ export async function handle(args: Result<ReturnType<typeof options>>) {
 
             // Scan the directory for candidates
-            env.DEBUG && console.time('[@tailwindcss/cli] Scan for candidates')
+            DEBUG && I.start('Scan for candidates')
```
These are called _after_ the initial `I.start('[@tailwincss/cli]')`, this allows us to remove the prefix because these logs will be indented.

RobinMalfait commented on packages/@tailwindcss-postcss/src/index.ts:
```diff
@@ -62,27 +64,54 @@ function tailwindcss(opts: PluginOptions = {}): AcceptedPlugin {
+          DEBUG && I.start(`[@tailwindcss/postcss] ${relative(base, inputFile)}`)
+
+          DEBUG && I.start('Quick bail check')
+          let canBail = true
+          root.walkAtRules((node) => {
+            if (
+              node.name === 'import' ||
+              node.name === 'theme' ||
+              node.name === 'config' ||
+              node.name === 'plugin' ||
+              node.name === 'apply'
+            ) {
+              canBail = false
+              return false
+            }
+          })
+          if (canBail) return
```
This is the quick bail where we are sure that Tailwind CSS should not be used. This prevents creating the compiler entirely.

philipp-spiess replied:
Should this be in a separate commit?

RobinMalfait replied:
I think you are right, it started this way and then diverged into 2 PRs. Created a separate one here: https://github.com/tailwindlabs/tailwindcss/pull/15340

RobinMalfait commented on packages/@tailwindcss-postcss/src/index.ts:
```diff
@@ -207,34 +241,43 @@ function tailwindcss(opts: PluginOptions = {}): AcceptedPlugin {
-                result.opts,
-              )
-              env.DEBUG && console.timeEnd('[@tailwindcss/postcss] Optimize CSS')
+              DEBUG && I.start('Optimization')
+
+              DEBUG && I.start('AST -> CSS')
+              let css = toCss(tailwindCssAst)
+              DEBUG && I.end('AST -> CSS')
+
+              DEBUG && I.start('Lightning CSS')
+              let ast = optimizeCss(css, {
+                minify: typeof optimize === 'object' ? optimize.minify : true,
+              })
+              DEBUG && I.end('Lightning CSS')
+
+              DEBUG && I.start('CSS -> PostCSS AST')
+              context.optimizedPostCssAst = postcss.parse(ast, result.opts)
+              DEBUG && I.end('CSS -> PostCSS AST')
+
+              DEBUG && I.end('Optimization')
```
Same as before, but split up explicitly

RobinMalfait commented on packages/tsconfig.base.json:
```diff
@@ -1,7 +1,8 @@
   "compilerOptions": {
-    "lib": ["ESNext", "DOM"],
-    "target": "ESNext",
+    "target": "es2022",
+    "lib": ["es2022", "esnext.disposable", "dom"],
```
These are needed to compile the new `using` keyword

RobinMalfait committed:
update changelog

RobinMalfait commented on CHANGELOG.md:
```diff
@@ -9,7 +9,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- Skip creating a compiler for CSS files that should not be processed ([#15303](https://github.com/tailwindlabs/tailwindcss/pull/15303))
 - Ensure `@import "…" reference` never generates utilities ([#15307](https://github.com/tailwindlabs/tailwindcss/pull/15307))
+- Improve debug logs to get better insights ([#15303](https://github.com/tailwindlabs/tailwindcss/pull/15303))
```
```suggestion
- Ensure `@import "…" reference` never generates utilities ([#15307](https://github.com/tailwindlabs/tailwindcss/pull/15307))
- Skip creating a compiler for CSS files that should not be processed ([#15303](https://github.com/tailwindlabs/tailwindcss/pull/15303))
- Improve debug logs to get better insights ([#15303](https://github.com/tailwindlabs/tailwindcss/pull/15303))
```

RobinMalfait committed:
Merge branch 'next' into feat/improve-debug-logs

RobinMalfait committed:
add debug logs for the quick bail check

RobinMalfait committed:
reset after reporting

RobinMalfait committed:
Merge branch 'next' into feat/improve-debug-logs

RobinMalfait committed:
remove explicit `white()` calls

RobinMalfait committed:
use `process.stderr.write`

thecrypticace reviewed (approved):
yay
